implementations/rust/ockam/ockam_node/Cargo.toml

```diff
@@ -36,6 +36,7 @@ tracing-subscriber = { version = "0.3", features = ["fmt", "env-filter"], option
 heapless = { version = "0.7", features = [ "mpmc_large" ], optional = true }
 ockam_node_no_std = { path = "../ockam_node_no_std", version = "0.12.0" , default-features = false, optional = true }
 ockam_executor = { path = "../ockam_executor", version = "^0.8.0", default-features = false, optional = true }
+async-trait = "0.1.42"
```

Comment:
ockam_core exports that dependency

Reply:
Hm, it shouldn't. We really should not be reexporting dependencies, it's a bad practice. But I can use it for now.

Reply:
@thomcc Could you please elaborate on why it's a bad practice?

Reply:
It increases the surface of our public APIs to include every API in the re-exported library. This means that if they have a major breaking change, we are forced to follow as well, or violate semver (in a very real way — note that we had an issue around another library doing this very thing in the past).

This is not worth the small convenience it adds.

Reply:
Note that it's okay to reexport dependencies when you control both. It can also be okay if you are okay with that tradeoff, but for this I don't think it's worth it. Ideally, we would even prevent `#[ockam::worker]` from being applied to non-worker implementations.

But it also means that you shouldn't release a version 1.0 stable without all of your public dependencies being there too.

Reply:
@thomcc we can create another typedef for async_trait to have more control over it (to be able to redefine it in the future if needed). Other than that, I don't understand what are the concerns about public API surface. async_trait directly affects all of our public traits (I think all of them are public). Correct me if I'm wrong, let's imagine user created their own implementation of our trait in their crate and declared async trait dependency directly, now if there's breaking change in async trait crate and ockam and user's crates have different version of it - it won't work. While if user used reexported async_trait dependency on their implementation - it would still work (worst case - there will be 2 async_trait crates in the final binary/library)

Reply:
> Correct me if I'm wrong, let's imagine user created their own implementation of our trait in their crate and declared async trait dependency directly, now if there's breaking change in async trait crate and ockam and user's crates have different version of it - it won't work

It depends on what the change is. In general, I suspect most changes to async_trait would still work here, but yes, it's possible that some could exist where it would not.

A bigger issue is users who use ockam without needing to write any custom workers, or otherwise interact with our async_trait export. Now because we reexport it, if it has a breaking change (say 0.3 to 0.4 is done to change), Ockam is forced to take a breaking change as well — this could be as significant as 1.0 to 2.0, which sounds major, but in reality would just because a minor part of our API surface had a breaking change. It would be unfortunate.

It's worth noting that the author of async_trait is generally quite good about marking crates that won't suffer any more breaking changes as 1.0, so I would not be that confident about async_trait being good for public usage here.

Reply:
@thomcc I don't think it's possible to write anything meaningful with ockam without creating your own workers... And making users add async_trait to their Cargo.toml along with ockam is not something we want as well

Reply:
> I don't think it's possible to write anything meaningful with ockam without creating your own workers

That's definitely distinct from what @mrinalwadhwa has indicated to me, but it does seem true at the moment.

Anyway, my long term plan for solving this issue with the `#[worker]` macro is to move it into our code, and make it specific to defining workers. As it is, someone can do:

```rs
#[ockam::worker]
trait SomeAsyncTraitThatHasNothingToDoWithWorkers { ... }

// or

#[ockam::worker]
impl SomeOtherAsyncTrait for MyType { ... }
```

and have it work. That's not something we want — you should only be allowed to use that macro with workers — most ways we might want to evolve the `worker` macro would be incompatble with it (for example, I've wanted it to automatically insert `tracing` calls — perhaps behind a feature), and would make adding such functionality into a breaking change when it doesn't need to be.

> making users add async_trait to their Cargo.toml along with ockam is not something we want as well

While I think where possible we should avoid having multiple Cargo.toml lines for most usage, honestly, that's a smaller concern than having API stability, IMO.

You'll note that even with `tokio`, their stability commitment has forced them to pull `tokio-streams` out, becuase they expect it to break sooner, and users of tokio that need stream functionality are forced to import both crates separately.

I don't think we can realistically be in the position where people using our APIs don't have to pull in any other libraries. We should just move it off of the critical path (as the note above about the `#[worker]` trait likely addresses it for defining worker).